### PR TITLE
Support for quantile regression

### PR DESCRIPTION
**Pull request by berhsuda**
Implemented support for quantile regression.  
Specifying 
objective = reg:quantile:0.60  sets the learner to try to predict the 60th quantile, and defaults the evaluation metric to 60th percentile quantile loss

As part of this, I implemented  special tree updaters based on https://statweb.stanford.edu/~jhf/ftp/trebst.pdf  for quantile regression.

### CODE REVIEW

**Comment by tqchen** [comment]
One side note, is there any reason why a special tree builder is needed? Normally gradient boosting procedure is totally isolated, so as long as we can take gradient, the original tree builder should work

**Comment by berhsuda** [comment]
One problem with quantile regression is that it has a zero second order
derivative of the loss function almost everywhere so the generic taylor
expansion approximation you use to derive the leaf score formula.

Also tried a differentiable approximation of the quantile regression
objective (  pseudo-huber loss) and that exhibits  really really slow
convergence.

-Dave

**Comment by tqchen** [comment]
What we actually need for convergence is not Taylor expansion, but a quadratic upper bound on the loss function(Taylor was a tight one). What we can do try is simply place 1 in turns of second order gradient, which should work given small enough learning rate

**Comment by berhsuda** [comment]
I tried this option as well.  The specialized updaters converge to a good
solution  more quickly than using the original tree builder.

I'll illustrate this with the performance of 3 different gbtrees on one of
my datasets.
The specialized tree updater after 10 iterations and shrinkage = 0.1 has an
error of: 10.86
The original tree builder with the approx you suggested and shrinkage = 0.1
:  16.11
The original tree builder with the approx you suggested and shrinkage = 0.5
:  15.23

The main problem is that in general the gradient is constant no matter how
far the prediction is from the label.  So updating the tree leaf scores
based on averaging the gradient essentially means that no prediction will
change by more than shrinkage * MAX(quantile,1-quantile) in each time
interval.

**Comment by tqchen** [comment]
Sorry for the long delayed respond on this. I did not merge this in mainly due to the delay in recent major refactor of code to restructure things.  This also enables quick plugin of these solvers easy.

See https://github.com/dmlc/xgboost/tree/brick/plugin

Since the current version relies on storing all the gradients in the statistics. this might not be as efficient, but you can always plug it in now into your environment and submit it back as a plugin first before we think how quantile regression can be efficiently done.
